apps/passport-client/components/screens/ImportBackupScreen.tsx

```diff
@@ -0,0 +1,386 @@
+import { PCDCollection } from "@pcd/pcd-collection";
+import { PCD } from "@pcd/pcd-types";
+import { SemaphoreGroupPCDTypeName } from "@pcd/semaphore-group-pcd";
+import { SemaphoreIdentityPCDTypeName } from "@pcd/semaphore-identity-pcd";
+import { useCallback, useEffect, useState } from "react";
+import { useFilePicker } from "use-file-picker";
+import { useDispatch, usePCDCollection } from "../../src/appHooks";
+import { getPackages } from "../../src/pcdPackages";
+import { AppState } from "../../src/state";
+import { useSelector } from "../../src/subscribe";
+import { Button, H2 } from "../core";
+import { MaybeModal } from "../modals/Modal";
+import { AppContainer } from "../shared/AppContainer";
+import { ScreenNavigation } from "../shared/ScreenNavigation";
+
+export function useImportScreenData() {
+  return useSelector<AppState["importScreen"]>((s) => s.importScreen, []);
+}
+
+const NoFolderSymbol = Symbol("None");
+
+// There are four main UI states that can occur after a user selects a file
+// to import.
+type ImportState =
+  // Initial state
+  | { state: "initial" }
+  // The selected file is valid, and the user can decide whether to import it.
+  | {
+      state: "valid-file-selected";
+      // The collection parsed from the selected file
+      collection: PCDCollection;
+      // The PCD IDs referring to PCDs which are valid to import
+      mergeablePcdIds: Set<PCD["id"]>;
+      // The folders the user has selected in the UI
+      selectedFolders: Set<string | symbol>;
+      // The PCD IDs that are valid and within the selected folders
+      selectedPcdIds: Set<PCD["id"]>;
+      // The count of valid PCDs in each importable folder
+      folderCounts: Record<string | symbol, number>;
+    }
+  // The import has been carried out, and `added` is the number of PCDs
+  // imported.
+  | { state: "import-complete" }
+  // The selected file is not valid.
+  | { state: "invalid-file" };
+
+export function ImportBackupScreen() {
+  const [importState, setImportState] = useState<ImportState>({
+    state: "initial"
+  });
+
+  // Global application state, used to report the success or failure of the
+  // import.
+  const importScreenState = useImportScreenData();
+
+  const { openFilePicker, filesContent } = useFilePicker({
+    accept: ".json",
+    multiple: false,
+    onFilesSelected: () => {
+      setImportState({ state: "initial" });
+    }
+  });
+
+  const existingPcdCollection = usePCDCollection();
+  const dispatch = useDispatch();
+
+  // Called when a valid file has been selected, and the user chooses to import
+  // PCDs from it.
+  const importPCDs = useCallback(() => {
+    // Should never happen, but makes TypeScript happy that we checked for it
+    if (importState.state !== "valid-file-selected") return;
+
+    dispatch({
+      type: "merge-import",
+      collection: importState.collection,
+      pcdsToMergeIds: importState.selectedPcdIds
+    });
+
+    setImportState({ state: "import-complete" });
```

Comment:
Given there's a possible failure case, and errors are being fed back from dispatch, I think there should be another "import-in-progress" state here, followed by "import-complete" only after the dispatch is done.

Reply:
The dispatch is technically asynchronous, but in practice is almost instantaneous. Since promises use the 'microtask' queue, it's quite likely that it completes before the next animation frame shows up. The only thing we could do with an intermediate state is show a "loading" message, but nobody would ever see it.